### PR TITLE
Transfer Hook Example with PDAs

### PR DESCRIPTION
**Pull request by buffalojoec**
This PR modifies the transfer hook example program to demonstrate use of PDAs as extra account configs

Ref #4942

### CODE REVIEW

**Review by joncinque — COMMENTED** [review]
Very cool to see this in action!

For an easier downstream dev experience, what do you think about refactoring the resolution bit from `ExtraAccountMeta::resolve` to take in the instruction components separately? So rather than taking `Instruction`, it takes the instruction data, pubkeys, and program id.

That way, an on-chain program can re-resolve to make sure all the inputs are correct. And then we won't have to hard-code anything in the example program, and people can copy it right away.

**Comment by buffalojoec** [comment]
> Very cool to see this in action!
> 
> 
> 
> For an easier downstream dev experience, what do you think about refactoring the resolution bit from `ExtraAccountMeta::resolve` to take in the instruction components separately? So rather than taking `Instruction`, it takes the instruction data, pubkeys, and program id.
> 
> 
> 
> That way, an on-chain program can re-resolve to make sure all the inputs are correct. And then we won't have to hard-code anything in the example program, and people can copy it right away.

Ahhh you're too quick!!

Yeah, I was just looking into this right now after seeing the test failure for `program-2022-test`. 

I’m curious your thoughts, but sounds like we are mostly in agreement. 

I was kind of thinking:
- On one hand, we shouldn't need to loop over TLV state to validate the provided extra metas, since with any on-chain program that requires them, it's much more likely they'd have a specific use for these accounts, therefore a specific check and/or business logic for each
- On the other, some form of client libraries may want to easily resolve TLV state without the specific use of an `Instruction`, so we want some extra sugar there
- I could be dead wrong on point #1

**Comment by joncinque** [comment]
I think you're right on both points -- while people won't want to loop over their accounts, they'll probably want an easy way to check that each one is the expected address without learning about all the contained types in tlv-account-resolution.

We could even include a `check_addresses` call which takes the raw slices to start with, and then they can extract them in the order expected. This way, they can also totally opt out of doing that and validate each address themselves. What do you think?

**Comment by buffalojoec** [comment]
> I think you're right on both points -- while people won't want to loop over their accounts, they'll probably want an easy way to check that each one is the expected address without learning about all the contained types in tlv-account-resolution.
> 
> We could even include a `check_addresses` call which takes the raw slices to start with, and then they can extract them in the order expected. This way, they can also totally opt out of doing that and validate each address themselves. What do you think?

OK I see where you're coming from. I guess the weird part corresponds to your original point of busting `Instruction` into components. Let me see what I can whip up

**Comment by buffalojoec** [comment]
Perhaps that will do it!

File: libraries/tlv-account-resolution/src/state.rs [hunk 1]
```diff
@@ -105,6 +143,35 @@ impl ExtraAccountMetaList {
...
+            .map(account_meta_from_info)
+            .collect::<Vec<_>>();
+
+        for config in extra_account_metas.iter() {
+            let meta = config.resolve(&provided_metas, instruction_data, program_id)?;
+            provided_metas
+                .iter()
+                .find(|&x| *x == meta)
+                .ok_or(AccountResolutionError::IncorrectAccount)?;
+        }
```
Comment (joncinque) [inline libraries/tlv-account-resolution/src/state.rs:170]: Is it also possible to ensure the order is correct? I think if you pass in `account_info_iter.as_slice()` you should also be able to enumerate and make sure the order is respected. And can you add a test that correctly errors if they're provided out of order?

**Review by joncinque — COMMENTED** [review]
Just one last little bit, then this should be good to go!

File: libraries/tlv-account-resolution/src/state.rs (hunk 1 above)
Comment (buffalojoec) [reply libraries/tlv-account-resolution/src/state.rs:170]: Sure thing!

**Review by joncinque — APPROVED** [review]
Looks great!

**Pull request merged**